docs/developer-tools/Guard.md

```diff
@@ -0,0 +1,104 @@
+    }
+
+    if (index < 0 || index >= array.Length)
+    {
+        throw new ArgumentOutOfRangeException(nameof(index), $"The index must be in the [0, {array.Length}) range, was {index}");
+    }
+
+    if (span.Length < array.Length)
+    {
+        throw new ArgumentException($"The target span is shorter than the input array, had a length of {span.Length}", nameof(span));
+    }
+
+    if (string.IsNullOrEmpty(text))
+    {
+        throw new ArgumentException("The input text can't be null or empty", nameof(text));
+    }
+}
+```
+
+And here is the same method, but using the new `Guard.APIs` to validate the input arguments:
+
+```csharp
+public static void SampleMethod(int[] array, int index, Span<int> span, string text)
+{
+    Guard.IsNotNull(array, nameof(array));
+    Guard.HasSizeAtLeast(array, 10, nameof(array));
+    Guard.IsInRangeFor(index, array, nameof(index));
+    Guard.HasSizeLessThanOrEqualTo(array, span, nameof(span));
+    Guard.IsNotNullOrEmpty(text, nameof(text));
+}
+```
+
+The `Guard` APIs will perform the required checks in the fastest way possible, and will throw the appropriate exception with a well formated message if they fail.
+
+## Methods
+
+There are dozens of different APIs and overloads in the `Guard` class, here are a few of them:
+
+| Methods | Return Type | Description |
+| -- | -- | -- |
+| IsNotNull<T>(T?, string) | void | Asserts that the input value is not null |
+| IsOfType<T>(object, string) | void | Asserts that the input value is of a specific type |
+| IsReferenceEqualTo<T>(T, T, string) | void | Asserts that the input value must be the same instance as the target value |
+| IsTrue(bool, string) | void | Asserts that the input value must be true |
+| IsEqualTo<T>(T, T, string) | void | Asserts that the input value must be equal to a specified value |
+| IsLessThan<T>(T, T, string) | void | Asserts that the input value must be less than a specified value |
+| IsLessThanOrEqualTo<T>(T, T, string) | void | Asserts that the input value must be less than or equal to a specified value |
+| IsInRange<T>(T, T, T, string) | void | Asserts that the input value must be in a given range |
+| IsInRangeFor<T>(int, T[], string) | void | Asserts that the input index is valid for a given array |
+| IsNotEmpty<T>(T[], string) | void | Asserts that the input array must not be empty |
+| HasSizeEqualTo<T>(T[], int, string) | void | Asserts that the input array must have a size of a specified value |
+| IsCompleted(Task, string) | void | Asserts that the input task is in a completed state |
+| IsNotCanceled(Task, string) | void | Asserts that the input task is not canceled |
+
+## Sample Code
+
+You can find more examples in our [unit tests](https://github.com/Microsoft/WindowsCommunityToolkit//blob/master/UnitTests/Diagnostics/Test_Guard.cs)
+
+## Requirements
+
+| Device family | Universal, 10.0.16299.0 or higher |
```

Comment:
These are .NET Standard, eh? There should be a line below here that states that, see example: https://docs.microsoft.com/en-us/windows/communitytoolkit/extensions/arrayextensions

Though wondering if we should clean this up or replace the 'Universal build number' in the grid with just '.NET Standard' as UWP supports that. Thoughts @vijay-nirmal?

Reply:
Done in https://github.com/MicrosoftDocs/WindowsCommunityToolkitDocs/pull/302/commits/78e2da312b3d4cf698f2b49475cb03ea11501d3b.

Yeah the `Guard` class only needs .NET Standard 2.1 - in fact I'd also like to multi target to .NET Standard 2.1 so that I can also sprinkle some `[NotNull]`, `[DoesNotReturnWhen]` etc. attributes on some of its APIs, which would help the analyzers when writing code using those APIs.

Reply:
@michael-hawker Do you mean something like we did in [Parser](https://docs.microsoft.com/en-us/windows/communitytoolkit/parsers/markdownparser)?

![image](https://user-images.githubusercontent.com/24755596/76430662-927c6b00-63d6-11ea-85de-2be9343ac60a.png)

I like it. I will do the cleanup work ✌

Reply:
@msvijayn That looks perfect! 😄

Reply:
@Sergio0694 You tagged the wrong person 🤣

Reply:
Whoops, I even double checked that!

I'm so sorry! 😅

Reply:
```suggestion
| Implementation | .NET Standard 2.0 |
```